pkg/controllers/nodeoverlay/validation/controller.go

```diff
@@ -0,0 +1,352 @@
+	}
+	for k, v := range nodeOverlay.Spec.Capacity {
+		s[key(nodePool, nodeOverlay, it)].Capacity[k] = v
+	}
+}
+
+func (s InstanceTypeOverlayStore) UpdateInstanceTypeOffering(nodePool v1.NodePool, nodeOverlay v1alpha1.NodeOverlay, it *cloudprovider.InstanceType, offering *cloudprovider.Offering) {
+	_, found := s[key(nodePool, nodeOverlay, it)]
+	overlayPriceChange := lo.Ternary(nodeOverlay.Spec.Price == nil, nodeOverlay.Spec.PriceAdjustment, nodeOverlay.Spec.Price)
+	if !found {
+		s[key(nodePool, nodeOverlay, it)] = &InstanceTypeUpdate{
+			instanceTypeName: it.Name,
+			nodePoolName:     nodePool.Name,
+			overlayName:      nodeOverlay.Name,
+			Offering:         map[string]*string{offering.Requirements.String(): overlayPriceChange},
+			Capacity:         corev1.ResourceList{},
+			weight:           nodeOverlay.Spec.Weight,
+		}
+	}
+	s[key(nodePool, nodeOverlay, it)].Offering[offering.Requirements.String()] = overlayPriceChange
+}
+
+func (s InstanceTypeOverlayStore) RemoveForUpdatesForOverlay(nodeOverlay v1alpha1.NodeOverlay) {
+	for k, v := range s {
+		if v.overlayName == nodeOverlay.Name {
+			delete(s, k)
+		}
+	}
+}
+
+func key(nodePool v1.NodePool, nodeOverlay v1alpha1.NodeOverlay, it *cloudprovider.InstanceType) string {
+	return fmt.Sprintf("%s-%s-%s", nodePool.Name, nodeOverlay.Name, it.Name)
+}
+
+func (s InstanceTypeOverlayStore) Reset() {
+	for key := range s {
+		delete(s, key)
+	}
+}
+
+// Controller for validating NodeOverlay configuration and surfacing conflicts to the user
+type Controller struct {
+	kubeClient               client.Client
+	cloudProvider            cloudprovider.CloudProvider
+	instanceTypeOverlayStore InstanceTypeOverlayStore
+}
+
+func (c *Controller) Name() string {
+	return "nodeoverlay.validation"
+}
+
+// NewController constructs a controller for node overlay validation
+func NewController(kubeClient client.Client, cp cloudprovider.CloudProvider, instanceTypeOverlayStore InstanceTypeOverlayStore) *Controller {
+	return &Controller{
+		kubeClient:               kubeClient,
+		cloudProvider:            cp,
+		instanceTypeOverlayStore: instanceTypeOverlayStore,
+	}
+}
+
+// Reconcile validates that all node overlays don't have conflicting requirements
+func (c *Controller) Reconcile(ctx context.Context, _ *v1alpha1.NodeOverlay) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	overlayList := &v1alpha1.NodeOverlayList{}
+	nodePoolList := &v1.NodePoolList{}
+	if err := c.kubeClient.List(ctx, overlayList); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeoverlays, %w", err)
+	}
+	err := c.kubeClient.List(ctx, nodePoolList)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodepool, %w", err)
+	}
+
+	overlayWithRuntimeValidationFailure := map[string]error{}
+	overlaysWithConflict := []string{}
+
+	overlayList.OrderByWeight()
+	for i := range overlayList.Items {
+		c.instanceTypeOverlayStore.RemoveForUpdatesForOverlay(overlayList.Items[i])
```

Comment:
When we're doing these updates, don't we want to make sure it's atomic, i.e. don't we want to ensure that the decorator doesn't read from the store once we're halfway through the pass? That way we always ensure that the full set of overlays have been validated together.